fix #6, add transaction api

manuzhang opened the pull request:
### This is not ready for merge but to gather comments as transaction api is a critical feature

There are a lot of changes so let me break down here to help review

Only two modifications are made to core

1. expose `upstreamMinClock` through `TaskContext` which is to decide when to checkpoint state 
2. add `withLong` and `getLong` to configure long value in user config

The rest is self contained in a new module `gearpump-experiment-state` under experiment. Half of them are UTs while the others are

1. API (`org.apache.gearpump.streaming.state.api`)

  * State
  * StateOp
  * Window
  * CheckpointManager
  * CheckpointStore

   I've added comments in each api. In brief, `State` is what user will use to manage states. Users could manipulate state by extending `StateOp`,  `State` is periodically checkpointed to a `CheckpointStore`, which is managed by a `CheckpointManager`. `Window` is defined for window function usages. 

2. Implementations (`org.apache.gearpump.streaming.state.impl`)

  Two state implementations are provided now (a SlidingWindowState will be added later)
  
     * DefaultState
     * FixedWindowState
   
  There is no need for users to write their own state management logic.

  A `DefaultCheckpointManager` is implemented. It maintains an internal checkpoint time, which decides when to checkpoint, and a pluggable `CheckpointStore`. 

  Only `InMemoryCheckpointStore` is available now and I'm working on a `HadoopStore`. 

3. Examples (`org.apache.gearpump.streaming.state.example`)

  There are two examples now.

    1. MessageCount

       a simple example to show how to use `DefaultState` to count the number of messages  received in all time.

    2.  WindowAverage

      a window application to show the usage of `FixedWindowState` to calculate the average of number received in each window.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/intel-hadoop/gearpump/837)
<!-- Reviewable:end -->

kkasravi commented:
Hi Tianlun

A few questions:

1. Could this be applied to the pipeline example? Right now pipeline aggregates messages (calculates a running average) until a certain time duration (as measured by message timestamps) has elapsed. It then writes to taskContext.output with this average value in a Message.msg. It seems like WindowAverage would be something very similar.

2. Could this be integrated within the high level api (experiments/dsl)? Maybe something like Stream[T].window(...)?

manuzhang commented:
Hi Kam. Yes,  for both questions.

1. PipeLine processors could make use of FixedWindowState which checkpoint and output state at the end of window. 
2. this is in plan.

kkasravi commented:
Tianlun - that's great - thanks!

BTW below is the build failure
```
[info] - DefaultCheckpointManager recover should read checkpoint at a time from store *** FAILED ***
[info]   MockitoException was thrown during property evaluation.
[info]     Message: 
[INFO] [04/18/2015 14:48:30.074] [MasterProxy@client526786716] sending identity to ActorSelection[Anchor(akka.tcp://master@127.0.0.1:59647/), Path(/user/master)]
[INFO] [04/18/2015 14:48:30.075] [MasterProxy@client526786716] sending identity to ActorSelection[Anchor(akka.tcp://master@127.0.0.1:59647/), Path(/user/master)]
[INFO] [04/18/2015 14:48:30.076] [MasterProxy@client1224904832] sending identity to ActorSelection[Anchor(akka.tcp://master@127.0.0.1:59647/), Path(/user/master)]
[INFO] [04/18/2015 14:48:30.076] [MasterProxy@client526786716] sending identity to ActorSelection[Anchor(akka.tcp://master@127.0.0.1:59647/), Path(/user/master)]
[info]   Mockito cannot mock this class: interface org.apache.gearpump.streaming.state.api.CheckpointStore
[info]   Mockito can only mock visible & non-final classes.
[info]   If you're not sure why you're getting this error, please report to the mailing list.
[info]     Occurred when passed generated values (
[info]       arg0 = 588,
[info]       arg1 = 554
[info]     )
[info] - DefaultCheckpointManager should close store *** FAILED ***
[info]   org.mockito.exceptions.base.MockitoException: Mockito cannot mock this class: interface org.apache.gearpump.streaming.state.api.CheckpointStore
[info] Mockito can only mock visible & non-final classes.
[info] If you're not sure why you're getting this error, please report to the mailing list.
[info]   at org.scalatest.mock.MockitoSugar$class.mock(MockitoSugar.scala:74)
[info]   at org.apache.gearpump.streaming.state.impl.DefaultCheckpointManagerSpec.mock(DefaultCheckpointManagerSpec.scala:30)
[info]   at org.apache.gearpump.streaming.state.impl.DefaultCheckpointManagerSpec$$anonfun$4.apply$mcV$sp(DefaultCheckpointManagerSpec.scala:71)
[info]   at org.apache.gearpump.streaming.state.impl.DefaultCheckpointManagerSpec$$anonfun$4.apply(DefaultCheckpointManagerSpec.scala:70)
[info]   at org.apache.gearpump.streaming.state.impl.DefaultCheckpointManagerSpec$$anonfun$4.apply(DefaultCheckpointManagerSpec.scala:70)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   ...
[info]   Cause: org.mockito.cglib.core.CodeGenerationException: java.lang.reflect.InvocationTargetException-->null
[info]   at org.mockito.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:238)[INFO] [04/18/2015 14:48:30.077] [MasterProxy@client1224904832] sending identity to ActorSelection[Anchor(akka.tcp://master@127.0.0.1:59647/), Path(/user/master)]
[info]   at org.mockito.cglib.proxy.Enhancer.createHelper(Enhancer.java:378)
[info]   at org.mockito.cglib.proxy.Enhancer.createClass(Enhancer.java:318)
[info]   at org.mockito.internal.creation.jmock.ClassImposterizer.createProxyClass(ClassImposterizer.java:124)
[info]   at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:60)
[info]   at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:52)
[info]   at org.mockito.internal.creation.CglibMockMaker.createMock(CglibMockMaker.java:24)
[info]   at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:32)
[info]   at org.mockito.internal.MockitoCore.mock(MockitoCore.java:59)
[info]   at org.mockito.Mockito.mock(Mockito.java:1258)
[info]   ...
[info]   Cause: java.lang.reflect.InvocationTargetException:
[info]   at sun.reflect.GeneratedMethodAccessor32.invoke(Unknown Source)
[info]   at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]   at java.lang.reflect.Method.invoke(Method.java:606)
[info]   at org.mockito.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:385)
[info]   at org.mockito.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:220)
[info]   at org.mockito.cglib.proxy.Enhancer.createHelper(Enhancer.java:378)
[info]   at org.mockito.cglib.proxy.Enhancer.createClass(Enhancer.java:318)
[info]   at org.mockito.internal.creation.jmock.ClassImposterizer.createProxyClass(ClassImposterizer.java:124)
[info]   at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:60)
[info]   at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:52)
[info]   ...
[info]   Cause: java.lang.OutOfMemoryError: PermGen space
[info]   at java.lang.ClassLoader.defineClass1(Native Method)
[info]   at java.lang.ClassLoader.defineClass(ClassLoader.java:800)
[info]   at sun.reflect.GeneratedMethodAccessor32.invoke(Unknown Source)
[info]   at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]   at java.lang.reflect.Method.invoke(Method.java:606)
[info]   at org.mockito.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:385)
[info]   at org.mockito.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:220)
[INFO] [04/18/2015 14:48:30.078] [MasterProxy@client526786716] sending identity to ActorSelection[Anchor(akka.tcp://master@127.0.0.1:59647/), Path(/user/master)]
[info]   at org.mockito.cglib.proxy.Enhancer.createHelper(Enhancer.java:378)
[info]   at org.mockito.cglib.proxy.Enhancer.createClass(Enhancer.java:318)
[info]   at org.mockito.internal.creation.jmock.ClassImposterizer.createProxyClass(ClassImposterizer.java:124)
[info]   ...
```

kkasravi commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/api/StateOp.scala:
```diff
@@ -0,0 +1,58 @@
+  def serialize(t: T): Array[Byte]
+
+  def deserialize(bytes: Array[Byte]): T
+}
```
We have a fair amount of serde exposed through out the codebase. Can we centralize or hide the exposure?

manuzhang replied:
Yes. We've got a few ways to serde although that still seems not enough (I had a hard time implementing TreeMap serde). The problem is that we can assume nothing for user data type so  this is necessary to persist data. Maybe we can provide one interface for all places where serde are required (source, sink, storage)

kkasravi commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/impl/FixedWindowState.scala:
```diff
@@ -0,0 +1,121 @@
+    if (states.isEmpty) None
+    else Option(states.maxBy(_._1)._1.endClock)
+  }
+}
```
For states and state transitions is there an advantage to using akka FSM rather than rolling our own?

manuzhang replied:
Actually, there is no state transitions here. State is a wrapper around data that will never be lost.

whjiang commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/api/CheckpointManager.scala:
```diff
@@ -0,0 +1,80 @@
+   * @param timestamp
+   * @param data
+   */
+  def checkpoint(timestamp: TimeStamp, data: Array[Byte]): Unit
```
Better to change `Array[Byte]` to `Iterator[Byte]`.

manuzhang replied:
any reason to use `Iterator` as we are not going iterate over data here ?

whjiang commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/api/CheckpointManager.scala:
```diff
@@ -0,0 +1,80 @@
+ *   1. an internal clock that is synced with an reliable external source
+ *   2. a checkpoint time which is advanced continuously by the configured
+ *   checkpoint interval on each checkpoint commit
+ */
```
Better add more explanation and some pseudo sample to breifly mention how this set of API is used.

whjiang commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/api/CheckpointManager.scala:
```diff
@@ -0,0 +1,80 @@
+   * @param timestamp
+   * @return
+   */
+  def recover(timestamp: TimeStamp): Option[Array[Byte]]
```
How long history the checkpoint store shall hold? Just the newest or latest 5 history, etc?

And what happens if fail to recover a timestamp?

manuzhang replied:
that should be configurable when we have the HadoopStore in. If the recovery fails, the init value defined by StateOp will be used.

whjiang commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/api/CheckpointManager.scala:
```diff
@@ -0,0 +1,80 @@
+ *   2. a checkpoint time which is advanced continuously by the configured
+ *   checkpoint interval on each checkpoint commit
+ */
+trait CheckpointManager {
```
Is this a central place for checkpoint or each task has its own checkpoint manager instance and perform checkpoint simultaneously and autonomic?

manuzhang replied:
each task would have its own checkpoint manager and perform checkpoint without interfering with others.

whjiang commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/example/WindowAverage.scala:
```diff
@@ -0,0 +1,64 @@
+    val userConfig = UserConfig.empty
+      .withValue[Duration](WindowCountProcessor.WINDOW_SIZE, windowSize)
+      // in a window application, checkpoint interval is set to window size such that
+      // every task will checkpoint state at the same pace
```
Do we really want this limitation? 

I remember Huawei mentioned they had use case that needs window size as 2 hour.

manuzhang replied:
It's simpler for now. I'll consider Huawei's case and provide a better design

whjiang commented:
General suggestions:
1. rename `State` to `StateHandler` and move it out of `api` package as it is NOT a user visible concept. 
2. rename `StateOp` to `State`.
3. add a bridge or proxy between `StateOp` and SummingBird Monoid to quickly support large set of states.
4. Create a subclass of `Task` as `WindowTask` or a trait and provide explicit API for setting window size.
5. Create a subclass of `Task` or a trait to provide explicit API for setting checkpoint info.

clockfly commented:
Hi Tianlun,

Several suggestions:
1. Seperate changes which are not directly related with transaction into a seperate PR. Like 
   minClock renamed to ```upstreamMinClock```

2. Prepare a demo and schedule a meeting to prove this transaction idea works.

manuzhang commented:
updated pr addressing the comments above with the changes:

1. rename `State` to `StateManager` (we already use `manager` in many places, just follow the convention)
2. rename `StateOp` to `State`.  `State` is a sealed trait. Users need extend its subclass `MonoidState` which incorporates Twitter Algebird for state update. `State` also uses Twitter Bijection for serialization / deserialization.
3. create a subclass of `Task`, `StatefulTask` which users extend to get state and transaction support 
4. create a class `Window` for user to set window parameters
5. add `HadoopCheckpointStore` (use `SequenceFile`)

manuzhang commented:
opened #914 to bring major updates
